Azure/sonic-buildimage
# [DHCP Relay]: Support Multiple VLANs (Separate DHCP Relay Agents, One Per VLAN)

**- What I did**
  - Spawn one DHCP relay agent for each VLAN that is configured as requiring one or more DHCP servers.
  - Also resolves https://github.com/Azure/sonic-buildimage/issues/985

**- How I did it**
  - docker-dhcp-relay entrypoint is a new script, docker_init.sh, which is responsible for generating the supervisord configuration file dynamically from a Jinja2 template. docker_init.sh then starts supervisord via exec so that supervisord still runs as PID 1
  - Dynamically-configured supervisord config now supports spawning a variable number of DHCP relay agent processes, based on VLAN configuration
  - Due to above changes, we no longer rely on /etc/default/isc-dhcp-relay config file, so the related Jinja2 template and shell script have been removed
 - Update isc-dhcp-relay patch such that it no longer relies on calling sonic-cfggen to retrieve port aliases with each packet (slow). Now parses the aliases from port_config.ini file directly once at startup and loads them into memory
 - Prevent relaying of already-relayed packets using `-m discard` parameter
- Change name of supervisord config file and template from generic `supervisord.conf` to more specific (albeit long) `docker-dhcp-relay.supervisord.conf` for better comprehension and also to avoid conflicts when copying to /usr/share/sonic/templates if we need to templatize other supervisord config files. We may also want to consider creating subdirectories under /usr/share/sonic/templates to further avoid other conflicts.
 - Update minigraph.py parser to parse out new `<DhcpRelays>` tag
 - Update sonic-cfggen test T0 minigraphs to contain new `<DhcpRelays>` tag
 - Remove executable permission and extra trailing newline from wait_for_intf.sh.j2 template file

**- How to verify it**
  1. Generate a T0 minigraph with two or more VLANs, each configured with one or more DHCP servers (using the `<DhcpRelays>` tag.
  2. Boot up SONiC using the minigraph created in step 1.
  3. Verify that there are multiple `dhcrelay` processes running, one per VLAN as configured in step 1.
  4. Verify that DHCP requests from devices on VLANs under the SONiC T0 are forwarded to all the DHCP servers configured in the minigraph, and verify all responses from DHCP servers are forwarded back to the appropriate device under the SONiC T0.

**- Description for the changelog**

[DHCP Relay]: Support Multiple VLANs (Separate DHCP Relay Agents, One Per VLAN)

**- A picture of a cute animal (not mandatory but encouraged)**
```
       _                        
       \`*-.                    
        )  _`-.                 
       .  : `. .                
       : _   '  \               
       ; *` _.   `*-._          
       `-.-'          `-.       
         ;       `       `.     
         :.       .        \    
         . \  .   :   .-'   .   
         '  `+.;  ;  '      :   
         :  '  |    ;       ;-. 
         ; '   : :`-:     _.`* ;
[bug] .*' /  .*' ; .*`- +'  `*' 
      `*-*   `*-*  `*-*'
```
